### ISSUE TITLE
Epoch level loss/metrics not being logged properly to logger (wandb)

### ISSUE THREAD

**Issue by snknitin** [Lightning-AI/lightning#18462]
### Bug description

This is an issue I've been facing since the new changes after migrating>2.0.0. The code below worked fine on 1.9.3 .this is the way I set it up to have a custom logging approach and I had to change few hooks like `on_training_epoch_end` to make the code work on 2.0.7, by following the migration PR from the changelog. 

Wandb is receiving every other metric except `train/epoch/.*` . train/step level loss and metrics and val+test /step and epoch level loss and metrics respectively are available to chart properly in the wandb logging charts. I'm not sure what argument combination of `self.log` I need to change.

*  Setting both `on_step` and `on_epoch`  to True raises a `NotImplemented Error`. 
*  Doing it the way i did here in the code gives me step level for all 3 and epoch level only for val and test. I'm not sure why this behavior is happening, 
* Tried to put everything on the `prog_bar` and set `logger=True` for everything. I know the rolling average is removed and the progress bar isn't auto populating the train epoch level metrics or even showing them when prog_bar is True. Only val/epoch shows up on progress bar. 

Has anyone observed this behavior or can correct me in the code , where i'm going wrong?  Here is the sampel code to reproduce it. you can add this to your LightningModule model.py 

 

```python
# compute
    def compute_metrics(self,preds,targets,mode):
        preds = preds.detach().cpu().int()
        targets = targets.detach().cpu().int()
        acc = self.metrics[mode](preds, targets)
        return acc

    # logging
    def logging_step(self,loss,acc,mode) -> None:
        self.log(f"{mode}/step/loss",loss,on_epoch=False ,on_step=True,prog_bar=True,logger=True)
        for metric_name, ac in acc.items():
            self.log(f"{mode}/step/{metric_name}",ac,on_epoch=False,on_step=True,prog_bar=True,logger=True)

    def logging_epoch(self, avg_loss,acc, mode):
        self.log(f"{mode}/epoch/loss", avg_loss, on_step=False,
                 on_epoch=True,
                 prog_bar=True,
                 logger=True)
        for metric_name,ac in acc.items():
            self.log(f"{mode}/epoch/{metric_name}", ac, on_step=False,
                     prog_bar=True,
                     logger=True,
                     on_epoch=True)
        self.metrics[mode].reset()

    def _shared_step(self, batch, batch_idx, mode):
        preds = self(batch.x_dict, batch.edge_index_dict, batch.edge_attr_dict) #.detach().cpu()
        targets = batch[self.label].edge_label #.detach().cpu()
        loss = self.loss_function(preds, targets)
        acc = self.compute_metrics(preds, targets, mode)
        self.logging_step(loss.item(), acc, mode)
        results = {'loss': loss.cpu(),'preds':preds.detach().cpu(),'targets':targets.detach().cpu()}
        return results

    def on_train_start(self):
        # by default lightning executes validation step sanity checks before training starts,
        # so we need to make sure val_acc_best doesn't store accuracy from these checks
        self.metrics["val"].reset()
        

    # Training
    def training_step(self, batch, batch_idx, *args, **kwargs) -> STEP_OUTPUT:
        mode = "train"
        results = self._shared_step(batch, batch_idx, mode)
        self.training_step_outputs.append(results)
        return results

    def validation_step(self,batch, batch_idx, *args, **kwargs) -> Optional[STEP_OUTPUT]:
        mode = "val"
        results = self._shared_step(batch, batch_idx, mode)
        self.validation_step_outputs.append(results)
        return results

    def test_step(self,batch, batch_idx, *args, **kwargs) -> Optional[STEP_OUTPUT]:
        mode = "test"
        results = self._shared_step(batch, batch_idx, mode)
        self.test_step_outputs.append(results)
        return results

    def on_training_epoch_end(self) -> None:
        mode="train"
        avg_loss = torch.stack([x["loss"] for x in self.training_step_outputs]).mean()
        all_preds = torch.vstack([x["preds"] for x in self.training_step_outputs])
        all_targets = torch.vstack([x["targets"] for x in self.training_step_outputs])
        self.training_step_outputs.clear()  # free memory
        acc = self.compute_metrics(all_preds,all_targets, mode)
        self.logging_epoch(avg_loss,acc, mode)
        print("\n")
        results = {'progress_bar': {'train_loss': avg_loss},
                   'train_loss': avg_loss}
        self.metrics[mode].reset()
        return results


    def on_validation_epoch_end(self) -> None:
        # [results,results,results ...]
        mode = "val"
        avg_val_loss = torch.stack([x["loss"] for x in self.validation_step_outputs]).mean()
        all_preds = torch.vstack([x["preds"] for x in self.validation_step_outputs])
        all_targets = torch.vstack([x["targets"] for x in self.validation_step_outputs])
        self.validation_step_outputs.clear()  # free memory
        acc = self.compute_metrics(all_preds, all_targets, mode)
        self.logging_epoch(avg_val_loss,acc, mode)
        results = {'progress_bar': {'val_loss':avg_val_loss},
                   'val_loss': avg_val_loss}
        self.metrics[mode].reset()
        return results

    def on_test_epoch_end(self):
        mode = "test"
        avg_test_loss = torch.stack([x["loss"] for x in self.test_step_outputs]).mean()
        all_preds = torch.vstack([x["preds"] for x in self.test_step_outputs])
        all_targets = torch.vstack([x["targets"] for x in self.test_step_outputs])
        self.test_step_outputs.clear()  # free memory
        acc = self.compute_metrics(all_preds, all_targets, mode)
        self.logging_epoch(avg_test_loss, acc, mode)
        results = {'progress_bar': {'test_loss': avg_test_loss},
                   'test_loss': avg_test_loss}
        #self.metrics["test"].reset()
        preds = all_preds.detach().cpu().numpy()
        targets = all_targets.detach().cpu().numpy()

        save_path = Path.cwd() / f"results_{self.hparams.run_dept}.npy"
        arr = np.hstack((preds.T, targets.T))
        np.save(save_path, arr)
        self.metrics[mode].reset()
        del arr
        return results
```

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
Sample code is in the description. Use it within a LightningModule
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

**Comment by ryan597** [comment]
I think you need to name the function as `on_training_epoch_end(self)`. See [LightningModule](https://lightning.ai/docs/pytorch/stable/common/lightning_module.html), but I think the validation one is fine anyways.

**Comment by snknitin** [comment]
yeah, that is what I have in the code snippet above. Which is why I'm confused because it works well for val and test but not for train.

**Comment by ryan597** [comment]
No you currently have the function as `on_training_epoch_end(self)`

**Comment by snknitin** [comment]
oh my god! I can't believe I missed that. So much time spent on such a tiny error. Thank you so much @ryan597
